File: sdks/typescript/trace-harness/src/feature/builtins.ts
import type { Node } from "../model/node";
import type { FeatureContext } from "./context";
import type { Feature } from "./feature";

function intervalUnion(intervals: Array<[number, number]>): number {
  if (intervals.length === 0) return 0;
  const sorted = [...intervals].sort((a, b) => a[0] - b[0]);
  let [start, end] = sorted[0]!;
  let total = 0;
  for (const [nextStart, nextEnd] of sorted.slice(1)) {
    if (nextStart <= end) {
      end = Math.max(end, nextEnd);
    } else {
      total += end - start;
      [start, end] = [nextStart, nextEnd];
    }
  }
  return total + end - start;
}

function selfMs(node: Node, context: FeatureContext): Record<string, unknown> {
  const children = context.children(node);
  if (children.length === 0) return {};
  const covered = intervalUnion(children.map((child) => [child.start_ms, child.end_ms]));
  return { self_ms: Math.round(Math.max(0, node.duration_ms - covered) * 1000) / 1000 };
}

function httpStatus(node: Node, context: FeatureContext): Record<string, unknown> {
  const statuses = context.children(node)
    .filter((child) => child.kind === "http")
    .map((child) => context.get(child, "status"))
    .filter((status): status is number => status !== undefined && status !== null)
    .map(Number);
  if (statuses.length === 0) return {};
  return { http_status: statuses.find((status) => status !== 200) ?? statuses[0] };
}

const FEATURES = [{ produces: ["self_ms"], applies: () => true, compute: selfMs, bake: true }, {
  produces: ["http_status"],
  applies: (node) => node.kind === "model-call",
  compute: httpStatus,
  bake: true,
}] satisfies Feature[];

export function builtinFeatures(): Feature[] {
  return [...FEATURES];
}

File: sdks/python/trace_harness/feature/builtins.py
"""harness 自带 Feature。零 biz；scoped TraceHarness 显式复制这组内置项。

都是 eager（bake=True）+ 读 facts/结构（不读 raw），故 IR-portable。
"""

from __future__ import annotations

from trace_harness.feature.ctx import Ctx
from trace_harness.feature.feature import Feature
from trace_harness.model.intervals import interval_union
from trace_harness.model.node import Node


def _self_ms(node: Node, ctx: Ctx) -> dict:
    """self_ms = 本节点 wall-clock duration − 子区间并集（未被子覆盖的独占耗时）。叶子不写。"""
    children = ctx.children(node)
    if not children:
        return {}
    covered = interval_union([(c.start_ms, c.end_ms) for c in children])
    return {"self_ms": round(max(0.0, node.duration_ms - covered), 3)}


def _http_status(node: Node, ctx: Ctx) -> dict:
    """model-call 从 http 子卷 http_status；非 200 优先（重试时成功那次会掩盖失败）。"""
    stats = [ctx.get(c, "status") for c in ctx.children(node) if c.kind == "http"]
    stats = [int(s) for s in stats if s is not None]
    if not stats:
        return {}
    return {"http_status": next((s for s in stats if s != 200), stats[0])}


BUILTIN_FEATURES = (
    Feature(("self_ms",), lambda n: True, _self_ms, bake=True),
    Feature(("http_status",), lambda n: n.kind == "model-call", _http_status, bake=True),
)
